[OMPIRBuilder] - Handle dependencies in `createTarget`

bhandarkar-pranav opened the pull request:
This PR handles dependencies specified by the `depend` clause on an OpenMP target construct. It does this much the same way clang does it by materializing an OpenMP `task` that is tagged with the dependencies.

To aid review, here are the relevant implementation details
1) `createTarget` - This function itself is largely unchanged except that it now accepts a vector of `DependData` objects that it simply forwards to `emitTargetCall`
2) `emitTargetCall` - This function has changed now to check if an outer target-task needs to be materialized (Yes, if target construct has `nowait` or has `depend` clause). If yes, it calls `emitTargetTask` to do all the heavy lifting for creating and dispatching the task.
3) `emitTargetTask` - Bulk of the PR is here. See the large comment explaining what it does at the beginning of the function.

Things that will/should follow in subsequent PRs
1) Handle this for target enter data, exit data and update data constructs
2) (Nice to have) - Outline logic for creating code for a task and use in `createTask` and `emitTargetTask`

bhandarkar-pranav committed:
Add a flag  to choose new codegen

bhandarkar-pranav committed:
clang prints for debugging

bhandarkar-pranav committed:
add an option -new-ompirbuilder-target-codegen to enable dependency-based target codegen path in OMPIRBuilder

bhandarkar-pranav committed:
checkpoint commit -> able to create an inlined version of the task that offloads

bhandarkar-pranav committed:
checkpoint commit

bhandarkar-pranav committed:
checkpoint commit

bhandarkar-pranav committed:
Simple test working. checkpoint commit. next steps clean up code, unittests(?) and lit tests - basically more testing before PR

bhandarkar-pranav committed:
clean up, clean up, everybody clean up

bhandarkar-pranav committed:
Add an MLIR lit test

bhandarkar-pranav committed:
add an end-to-end offloading test for target depend

bhandarkar-pranav committed:
Clean up, clean up, everybody clean up (some more)

bhandarkar-pranav committed:
Add back some ws that was removed

bhandarkar-pranav committed:
More cleanup and comments

bhandarkar-pranav committed:
clang-format fixes

bhandarkar-pranav committed:
remove commented out createFakeIntVal

bhandarkar-pranav committed:
more cleanup

bhandarkar-pranav committed:
Fix dependencies when nowait is used on target construct

bhandarkar-pranav committed:
Add comments for emitTargetTask and emitDepArray

bhandarkar-pranav committed:
Fix comment in target-depend.f90

bhandarkar-pranav commented:
ping! Gentle reminder, I'd appreciate reviews on this.

ergawy commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5229,13 +5362,288 @@ static void emitTargetOutlinedFunction(
+    SmallVector<llvm::OpenMPIRBuilder::DependData> &Dependencies,
+    bool HasNoWait) {
+
+  // When we arrive at this function, the target region itself has been
```
Really appreciate this block of comments. Paints the picture of what happens clearly. Thanks :)!

ergawy commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5229,13 +5362,288 @@ static void emitTargetOutlinedFunction(
+                                     RTLoc, TargetTaskAllocaIP));
+
+  OI.ExitBB = Builder.saveIP().getBlock();
+  OI.PostOutlineCB = [this, ToBeDeleted, Dependencies,
```
I there are pieces in this lambda that can be outlined into shared utils and called from here and `OpenMPIRBuilder::createTask`. For example, the logic to setup `TaskData` (but, other parts as well).

bhandarkar-pranav replied:
Yes, that is an active Todo which I was planning to add in a different PR because now this PR doesn't touch the handling of the `task` construct at all. But that's only a preference, I'll be happy to roll those changes into this PR too.

Meinersbur replied:
This passes `ToBeDeleted` and `Dependencies` by-value to the lambda, i.e. makes a copy of those lists. If called multiple times, items in `ToBeDeleted` will be erased multiple times. I assume this is not intended?
```suggestion
  OI.PostOutlineCB = [this, &ToBeDeleted, &Dependencies,
```

bhandarkar-pranav replied:
> This passes `ToBeDeleted` and `Dependencies` by-value to the lambda, i.e. makes a copy of those lists. If called multiple times, items in `ToBeDeleted` will be erased multiple times. I assume this is not intended?

In fact a copy is necessary to "own" the data. `PostOutlineCB` is called only once so multiple deletions is not a problem. However, and what is more important to note, is that it is called long after `ToBeDeleted` is destroyed (that is after we have returned from `emitTargetTask`) making it necessary to keep a copy around.

ergawy reviewed (commented):
Thanks @pranavb-ca! I did a first round and have a few comments.

skatrak commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5572,6 +5999,8 @@ void OpenMPIRBuilder::emitOffloadingArrays(
     return;
 
   Builder.restoreIP(AllocaIP);
+  LLVM_DEBUG(dbgs() << "Basicblock before emitOffloadingArrays\n"
+                    << *(Builder.GetInsertBlock()) << "\n");
```
Nit: Is this addition intentional?

bhandarkar-pranav replied:
No, this should go :(

skatrak reviewed (commented):
Thank you Pranav, I think this looks good. I just have a few minor comments.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -792,6 +792,9 @@ void OpenMPIRBuilder::finalize(Function *Fn) {
   if (!OffloadInfoManager.empty())
     createOffloadEntriesAndInfoMetadata(ErrorReportFn);
+
+  LLVM_DEBUG(dbgs() << "Module after OMPIRBuilder::finalize\n");
+  LLVM_DEBUG(dbgs() << M << "\n");
```
Sorry for having left these here. This should go.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+//   the corresponding dependence.
+// - All code is inserted in the entry block of the current function.
+static Value *
+emitDepArray(OpenMPIRBuilder &OMPBuilder,
```
Good point. I'll rename this function.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+  Module &M = OMPBuilder.M;
+
+  Value *DepArray = nullptr;
+  if (Dependencies.size()) {
```
Aah nice catch.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+    Type *DepArrayTy = ArrayType::get(DependInfo, Dependencies.size());
+    DepArray = Builder.CreateAlloca(DepArrayTy, nullptr, ".dep.arr.addr");
+
+    unsigned P = 0;
+    for (const OpenMPIRBuilder::DependData &Dep : Dependencies) {
```
Will do, thanks.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+    DepArray = Builder.CreateAlloca(DepArrayTy, nullptr, ".dep.arr.addr");
+
+    unsigned P = 0;
+    for (const OpenMPIRBuilder::DependData &Dep : Dependencies) {
```
Actually, my intention was to switch `OpenMPIRBuilder::createTask` to use this function but i had a couple of other changes to `OpenMPIRBuilder::createTask` in mind so I decided to leave that out and roll it into a different PR. If you prefer, I can make the switch in this PR itself though. Just a preference, no strong opinions on this.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+// void @.omp_target_task_proxy_func(i32 %thread.id, ptr %task)
+// This function is called from emitTargetTask once the
+// code to launch the target kernel has been outlined already.
+static Function *emitProxyTaskFunction(OpenMPIRBuilder &OMPBuilder,
```
Good point @ergawy and @skatrak I'll pick a different name `emitTargetTaskProxyFunction` sounds appropriate.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+  // CalledFunction is the target launch function, i.e.
+  // the function that sets up kernel arguments and calls
+  // __tgt_target_kernel to launch the kernel on the device.
+  Function *CalledFunction = StaleCI->getCalledFunction();
```
Yes, that's a better name.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+      BasicBlock::Create(Builder.getContext(), "entry", ProxyFn);
+  Builder.SetInsertPoint(EntryBB);
+
+  bool HasShareds = StaleCI->arg_size() > 1;
```
Got it,  yes. will do that.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+           "for extracted function");
+    StructType *ArgStructType =
+        dyn_cast<StructType>(ArgStructAlloca->getAllocatedType());
+    LLVM_DEBUG(dbgs() << "ArgStructType = " << *ArgStructType << "\n");
```
Yes, this should have been removed before opening the PR even. I accidentally left something I had put in place when i was debugging my work.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+
+    Builder.CreateCall(CalledFunction, {ThreadId, NewArgStructAlloca});
+  }
+  ProxyFn->getArg(0)->setName("thread.id");
+  ProxyFn->getArg(1)->setName("task");
```
Sure, will do.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5252,6 +5660,8 @@ static void emitTargetCall(OpenMPIRBuilder &OMPBuilder, IRBuilderBase &Builder,
   //  emitKernelLaunch
   auto &&EmitTargetCallFallbackCB =
       [&](OpenMPIRBuilder::InsertPointTy IP) -> OpenMPIRBuilder::InsertPointTy {
+    LLVM_DEBUG(dbgs() << "EmitTargetCallFallbackCB::Builder = " << &Builder
+                      << "\n");
```
No, this should go. sorry about this.

bhandarkar-pranav commented on offload/test/offloading/fortran/target-depend.f90:
```diff
@@ -0,0 +1,37 @@
+
+  z = 1
+  !$omp task depend(out: z) shared(z)
+  z = N
```
Yes, that makes sense. I'll add a omp parallel also so that if I remove the depend there is a chance that i'd get the wrong result. That'll really prove this works.

Meinersbur commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -2253,7 +2275,8 @@ class OpenMPIRBuilder {
                              TargetBodyGenCallbackTy BodyGenCB,
-                             TargetGenArgAccessorsCallbackTy ArgAccessorFuncCB);
+                             TargetGenArgAccessorsCallbackTy ArgAccessorFuncCB,
+                             SmallVector<DependData> Dependencies = {});
```
```suggestion
                             ArrayRef<DependData> Dependencies = {});
```

Meinersbur commented on llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:
```diff
@@ -1762,6 +1762,26 @@ class OpenMPIRBuilder {
+      Function *OutlinedFn, Value *OutlinedFnID,
+      EmitFallbackCallbackTy EmitTargetCallFallbackCB, TargetKernelArgs &Args,
+      Value *DeviceID, Value *RTLoc, InsertPointTy AllocaIP,
+      SmallVector<OpenMPIRBuilder::DependData> &Dependencies, bool HasNoWait);
```
```suggestion
      ArrayRef<OpenMPIRBuilder::DependData> Dependencies, bool HasNoWait);
```

bhandarkar-pranav replied:
I don't think this is safe. I capture Dependencies in a lambda so storing ArrayRef which doesnt manage the lifetime of the data pointer isn't safe.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+emitDepArray(OpenMPIRBuilder &OMPBuilder,
+             SmallVector<OpenMPIRBuilder::DependData> &Dependencies) {
+  // Early return if we have no dependencies to process
+  if (!Dependencies.size())
```
```suggestion
  if (Dependencies.empty())
```

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+  Value *DepArray = nullptr;
+  if (Dependencies.size()) {
+    OpenMPIRBuilder::InsertPointTy OldIP = Builder.saveIP();
+    Builder.SetInsertPoint(
+        &OldIP.getBlock()->getParent()->getEntryBlock().back());
```
```suggestion
    Builder.SetInsertPoint(
        &OldIP.getBlock()->getParent()->getEntryBlock().getTerminator());
```
just makes it more explicit. 
Did you consider explicitly passing the AllocaIP, could be needed with combined constructs.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -1698,6 +1701,64 @@ void OpenMPIRBuilder::createTaskyield(const LocationDescription &Loc) {
+
+    unsigned P = 0;
+    for (const OpenMPIRBuilder::DependData &Dep : Dependencies) {
+      Value *Base =
+          Builder.CreateConstInBoundsGEP2_64(DepArrayTy, DepArray, 0, P);
+      // Store the pointer to the variable
+      Value *Addr = Builder.CreateStructGEP(
+          DependInfo, Base,
+          static_cast<unsigned int>(RTLDependInfoFields::BaseAddr));
+      Value *DepValPtr =
+          Builder.CreatePtrToInt(Dep.DepVal, Builder.getInt64Ty());
+      Builder.CreateStore(DepValPtr, Addr);
+      // Store the size of the variable
+      Value *Size = Builder.CreateStructGEP(
+          DependInfo, Base,
+          static_cast<unsigned int>(RTLDependInfoFields::Len));
+      Builder.CreateStore(Builder.getInt64(M.getDataLayout().getTypeStoreSize(
+                              Dep.DepValueType)),
+                          Size);
+      // Store the dependency kind
+      Value *Flags = Builder.CreateStructGEP(
+          DependInfo, Base,
+          static_cast<unsigned int>(RTLDependInfoFields::Flags));
+      Builder.CreateStore(
+          ConstantInt::get(Builder.getInt8Ty(),
+                           static_cast<unsigned int>(Dep.DepKind)),
+          Flags);
```
Consider adding a comment with a mock source of what is generate here:
```
DepArray[P].BaseAddre = ...
```

Meinersbur reviewed (commented):
Not a full a review, but the first notes that I started with

Meinersbur commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -681,7 +681,30 @@ convertOmpTeams(omp::TeamsOp op, llvm::IRBuilderBase &builder,
   return bodyGenStatus;
 }
-
+static void
```
[nit]
```suggestion

static void
```

Meinersbur commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -681,7 +681,30 @@ convertOmpTeams(omp::TeamsOp op, llvm::IRBuilderBase &builder,
+static void
+buildDependData(std::optional<ArrayAttr> depends, OperandRange dependVars,
+                LLVM::ModuleTranslation &moduleTranslation,
+                SmallVector<llvm::OpenMPIRBuilder::DependData> &dds) {
```
```suggestion
                SmallVectorImpl<llvm::OpenMPIRBuilder::DependData> &dds) {
```

Meinersbur reviewed (commented):
Conceptually, looks quite good. Just some style comments.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5212,6 +5273,78 @@ static Function *createOutlinedFunction(
+    LoadInst *LoadShared =
+        Builder.CreateLoad(PointerType::getUnqual(Ctx), Shareds);
+
+    // TODO: Are these alignment values correct?
```
Thanks

ergawy reviewed (approved):
LGTM

bhandarkar-pranav committed:
Incorporate changes for review comments

bhandarkar-pranav committed:
Update testcase to make it more reliable by correcting the use of the depend clause (sibling tasks)

bhandarkar-pranav commented:
@Meinersbur, @ergawy, @skatrak  - Could you please give this another look. I have updated the testcase `target-depend.f90` because the previous version was flaky; I had used `depend` incorrectly which I have fixed now. (Previously, `depend` was used on tasks that were not sibling tasks)

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5272,36 +5682,53 @@ static void emitTargetCall(OpenMPIRBuilder &OMPBuilder, IRBuilderBase &Builder,
+  // The presence of certain clauses on the target directive require the
+  // explicit generation of the target task.
+  if (RequiresOuterTargetTask) {
+    OMPBuilder.emitTargetTask(OutlinedFn, OutlinedFnID,
```
Thanks for point this out. I want to use this opportunity to understand this idiom I see all around in `OMPIRBuilder.cpp` which is ``Builder.restoreIP(FunctThatReturnsInsertPointTy())``. Now, if the only return from `FuncThatReturnsInsertPointTy` is `Builder.saveIP()`, then do I need a call to `Builder.restoreIP()` on return ? Because `emitTargetTask()` returns `Builder.saveIP()`, which, in other words, means it is returning the state of the Builder's insert point. We then restore the same `Builder` to that insert point which sounds redundant. However, I see this in a number of places in `OMPIRBuilder.cpp`

skatrak replied:
I agree with you, since I have noticed this redundant pattern as well, though in a way I think it makes sense from a uniformity point of view. There are several codegen functions that return an insertion point, which should be used by the caller to proceed with any further codegen.

I think that the fact that oftentimes the insertion point that's being returned is indeed already the current one shouldn't be relied on by the caller unless we make this a contract followed by all such functions. To be consistent, I think these codegen function must either always return an insertion point which must be used by the caller (the current approach) or all codegen functions must not return an insertion point and set it themselves to the spot follow-up codegen should go.

In my opinion, I believe it's currently better to keep the current approach even though it results in some redundant setting of the insertion point. Skipping the call to `Builder.restoreIP()` in the caller could result in unintended problems in the future, since the current contract is that the returned insertion point is the one that should be used and not the current one.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -705,28 +728,9 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
-      dds.emplace_back(dd);
-    }
-  }
+  if (!taskOp.getDependVars().empty() && taskOp.getDepends())
```
It looks like this comment hasn't been addressed. Let me know if you just missed it, if you disagree with it or if it isn't very clear and I should try to explain better.

skatrak reviewed (commented):
Thanks for working on my previous comments. I just have a couple more minor ones.

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5229,13 +5382,284 @@ static void emitTargetOutlinedFunction(
+
+    Function *ProxyFn = emitTargetTaskProxyFunction(*this, Builder, StaleCI);
+
+    LLVM_DEBUG(dbgs() << "Proxy task entry function created: " << *ProxyFn
```
This one is intentional because as more work is done for `target task` (and more work remains) being able to see this function will be useful for debugging. At a later, more stable point, we can get rid of this. Just my preference to leave this around for now.

bhandarkar-pranav committed:
Address more review comments

bhandarkar-pranav commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3088,10 +3102,13 @@ convertOmpTarget(Operation &opInst, llvm::IRBuilderBase &builder,
       kernelInput.push_back(mapData.OriginalValue[i]);
   }
+  SmallVector<llvm::OpenMPIRBuilder::DependData> dds;
+  buildDependData(targetOp.getOperation(), moduleTranslation, dds);
```
@skatrak  - Apologies for missing your previous comment about moving the empty depends list check into `buildDependData`. I have done that now in this iteration of the PR.

bhandarkar-pranav committed:
Incorporate one more review comment

bhandarkar-pranav commented:
@skatrak - I have updated the PR with changes based on your feedback. Could you please take a look?

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -682,6 +682,41 @@ convertOmpTeams(omp::TeamsOp op, llvm::IRBuilderBase &builder,
 }
 
+static void
+buildDependData(Operation *op, LLVM::ModuleTranslation &moduleTranslation,
```
I'd suggest taking `std::optional<ArrayAttr> depends` and `OperandRange dependVars` as arguments instead of `Operation *op` here. This way you don't need to try casting `op` to each supported operation below, but rather make the body of this function what the `processDepends` lambda inside currently has.

skatrak commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -705,28 +740,7 @@ convertOmpTaskOp(omp::TaskOp taskOp, llvm::IRBuilderBase &builder,
-      dds.emplace_back(dd);
-    }
-  }
+  buildDependData(taskOp.getOperation(), moduleTranslation, dds);
```
Suggestion to how this call could look if following my comment above:
```suggestion
  buildDependData(moduleTranslation, taskOp.getDepends(), taskOp.getDependVars(), dds);
```

skatrak reviewed (approved):
Thank you Pranav, LGTM. I have just one more comment but no need to wait for another review by me before merging this. Just make sure @Meinersbur is happy with it as well!

bhandarkar-pranav committed:
Incorporate changes from review comments

bhandarkar-pranav commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5229,13 +5362,288 @@ static void emitTargetOutlinedFunction(
+  OI.OuterAllocaBB = AllocaIP.getBlock();
+
+  // Add the thread ID argument.
+  std::stack<Instruction *> ToBeDeleted;
```
@Meinersbur  -> The only reason, I didnt do it is because `createFakeIntVal` is used by a few other functions not related to this PR. So changing `ToBeDeleted` from `std::stack`  to `SmallVector` would involve changing other functions (such as `createTeams`) which are not related to this PR. I could go ahead and add that to this PR. Or simply open a separate PR just to make this change which should be quick. Let me know what you prefer.

bhandarkar-pranav replied:
@Meinersbur - Please take a look. I have gone ahead and made this change. If you agree with my reason above then I'll simply back out that one [commit](https://github.com/llvm/llvm-project/pull/93977/commits/a314a84964ae243cb4fee98a9adeacdfa9edf71e) and make it a new PR.

bhandarkar-pranav committed:
Simplify test to make it faster yet effective

bhandarkar-pranav committed:
Use SmallVector instead of std::stack in CreateFakeIntVal

bhandarkar-pranav committed:
Use SmallVectorImpl in emitTaskDependencies

Meinersbur reviewed (approved):
LGTM
